flo-runner/src/main/java/com/spotify/flo/context/FloRunner.java

```diff
@@ -20,116 +20,113 @@
+   * @param task task to run
+   * @param config configuration to apply
+   * @param <T> type of task
+   * @return a {@link Result} with the value and throwable (if thrown)
+   */
+  public static <T> Result<T> runTask(Task<T> task, Config config) {
+    return new FloRunner<T>(config).run(task);
   }
 
-  public static void runTask(TaskConstructor<?> taskConstructor, String... args) {
-    final Service service = Services.usingName(NAME)
-        .withEnvVarPrefix("FLO")
-        .build();
-
-    try (Service.Instance i = service.start(args)) {
-      final long t0 = System.currentTimeMillis();
-      final FloRunner runner = new FloRunner(i);
-      final Task<?> task = runner.run(taskConstructor);
-
-      i.waitForShutdown();
-      final long elapsed = System.currentTimeMillis() - t0;
-      logging.complete(task.id(), elapsed);
-    } catch (Throwable e) {
-      logging.exception(e);
-      System.exit(1);
-    }
-
-    System.exit(exitCode);
+  /**
+   * Run task and return {@link Result} containing the last value or throwable.
+   * @param task task to run
+   * @param <T> type of task
+   * @return a {@link Result} with the value and throwable (if thrown)
+   */
+  public static <T> Result<T> runTask(Task<T> task) {
+    return runTask(task, ConfigFactory.load("flo"));
   }
 
-  private Task<?> run(TaskConstructor<?> taskConstructor) {
-    logging = (isMode("tree") && isJsonMode()) ? new JsonTreeLogging() : logging;
+  private Result<T> run(Task<T> task) {
     logging.init();
-    closer.register(logging);
 
-    final ImmutableList<String> unprocessedArgs = instance.getUnprocessedArgs();
-    final String[] taskArgs = unprocessedArgs.toArray(new String[unprocessedArgs.size()]);
+    closeables.add(logging);
 
     logging.header();
 
-    final Task<?> task = taskConstructor.create(taskArgs);
-
     if (isMode("tree")) {
       logging.tree(TaskInfo.ofTask(task));
-      exit(0);
-    } else {
-      final EvalContext evalContext = createContext();
-      logging.printPlan(TaskInfo.ofTask(task));
-      final EvalContext.Value<?> value = evalContext.evaluate(task);
-      value.consume((ˍ) -> exit(0));
-      value.onFail(this::exit);
+      return new Result<>(CompletableFuture.completedFuture(null));
     }
 
-    return task;
+    logging.printPlan(TaskInfo.ofTask(task));
+
+    final EvalContext evalContext = createContext();
+    final long t0 = System.currentTimeMillis();
+    final EvalContext.Value<T> value = evalContext.evaluate(task);
+    final CompletableFuture<T> future = new CompletableFuture<>();
+
+    value.consume(v -> {
+      final long elapsed = System.currentTimeMillis() - t0;
+      logging.complete(task.id(), elapsed);
+      future.complete(v);
+      try {
+        close();
+      } catch (Exception e) {
+        e.printStackTrace();
+      }
+    });
+
+    value.onFail(t -> {
+      logging.exception(t);
+      final long elapsed = System.currentTimeMillis() - t0;
```

Comment:
This code is identical to the `consume` handler above… I couldn't find a way to handle it in one place :/

Reply:
found a way :)